File: cases/taint/php/recursive-carry-positive/recursive_carry_positive.php
<?php
function dfb_source(): string { // DFB-SOURCE: recursive-carry-input
    return "tainted";
}

function dfb_sink(string $value): void {} // DFB-SINK: recursive-carry-sink

function carry(string $value, int $depth): string { // DFB-WITNESS: recursive-carry-step
    if ($depth === 0) {
        return $value;
    }
    return carry($value, $depth - 1);
}

function run(): void {
    dfb_sink(carry(dfb_source(), 5));
}

File: cases/taint/java/recursive-carry-positive/RecursiveCarryPositive.java
package dataflowbench.taint;

final class RecursiveCarryPositive {
    static int dfb_source() { // DFB-SOURCE: recursive-carry-input
        return 1;
    }

    static void dfb_sink(int value) { } // DFB-SINK: recursive-carry-sink

    static int carry(int value, int depth) { // DFB-WITNESS: recursive-carry-step
        if (depth == 0) {
            return value; // DFB-WITNESS: recursive-carry-base
        }
        return carry(value, depth - 1);
    }

    static void run() {
        dfb_sink(carry(dfb_source(), 5));
    }
}
